### PR TITLE
Support MSC1767's `content.body` behaviour; Add base rules from MSC3933

### PR DESCRIPTION
**Pull request by turt2live**
## Notes for reviewer
**This is reviewable commit-by-commit (recommended).**

**It may be best to use the compare view instead of github for this: https://github.com/matrix-org/synapse/compare/pull/14521/head...pull/14524/head** (this PR is based on https://github.com/matrix-org/synapse/pull/14521)

## PR series

This PR is part of a series:
* https://github.com/matrix-org/synapse/pull/14520
* https://github.com/matrix-org/synapse/pull/14521
* https://github.com/matrix-org/synapse/pull/14524 - You are here

### CODE REVIEW

File: rust/src/push/base_rules.rs [hunk 1]
```diff
@@ -274,6 +274,156 @@ pub const BASE_APPEND_UNDERRIDE_RULES: &[PushRule] = &[
         default: true,
         default_enabled: true,
     },
+    PushRule {
+        rule_id: Cow::Borrowed(
+            "global/underride/.org.matrix.msc3933.rule.extensible.encrypted_room_one_to_one",
+        ),
+        priority_class: 1,
+        conditions: Cow::Borrowed(&[
+            Condition::Known(KnownCondition::EventMatch(EventMatchCondition {
```
Comment (MatMaul) [inline rust/src/push/base_rules.rs:283]: Regarding conditions order, does it matter (I am a bit new to push rules) ?
They are not in the same order as in the MSC. Did you tweak the order for performance reason for example ?

Comment (turt2live) [reply rust/src/push/base_rules.rs:283]: I don't think it matters, as the whole array is effectively a giant `if` statement: if any of the conditions fail, the push rule's execution fails, so it shouldn't matter if the condition "fails" first or last.

Comment (MatMaul) [reply rust/src/push/base_rules.rs:283]: Ok so just pointing that perhaps we can improve perfs by trying to order the conditions from most likely to fail to least likely.
The `likely` part will however change quite drastically between right after deploy (almost no extensible compatible rooms) and a year later when a lot of rooms would have migrate/be created with the new room version.
And this is probably over optimization I guess, so I would say let's keep it like that and revisit if it happens to be a hot code path later on.

**Review by MatMaul — APPROVED** [review]

**Comment by turt2live** [comment]
rebasing to account for stacked PRs, sorry.

**Comment by turt2live** [comment]
I think this PR's tests are failing for unrelated reasons, namely tests added/modified in https://github.com/matrix-org/synapse/pull/14471

force merging to land anyways, and will monitor develop build.

**Pull request merged**